crates/sui-framework/packages/deepbook/sources/math.move

```diff
@@ -6,25 +6,123 @@ module deepbook::math {
         let y = (y as u128);
         let is_round_down = true;
         if ((x * y) % FLOAT_SCALING_U128 == 0) is_round_down = false;
         (is_round_down, ((x * y / FLOAT_SCALING_U128) as u64))
     }
 
-    public fun div_round(x: u64, y: u64): (bool, u64) {
+    // multiply two floating numbers and assert the result is non zero
+    // also returns whether the result is rounded down
+    public fun safe_mul_round(x: u64, y: u64): (bool, u64) {
+        let (is_round_down, result) = unsafe_mul_round(x, y);
+        assert!(result > 0, EUnderflow);
+        (is_round_down, result)
+    }
+
+    // divide two floating numbers
+    // also returns whether the result is rounded down
+    public fun unsafe_div_round(x: u64, y: u64): (bool, u64) {
         let x = (x as u128);
         let y = (y as u128);
         let is_round_down = true;
         if ((x * (FLOAT_SCALING as u128) % y) == 0) is_round_down = false;
         (is_round_down, ((x * (FLOAT_SCALING as u128) / y) as u64))
     }
+
+    // divide two floating numbers and assert the result is non zero
+    // also returns whether the result is rounded down
+    public fun safe_div_round(x: u64, y: u64): (bool, u64) {
+        let (is_round_down, result) = unsafe_div_round(x, y);
+        assert!(result > 0, EUnderflow);
+        (is_round_down, result)
+    }
+
+    #[test_only] use sui::test_utils::assert_eq;
+    #[test]
+    fun test_mul() {
+        assert_eq(unsafe_mul(1_000_000_000, 1), 1);
+        assert_eq(unsafe_mul(9_999_999_999, 1), 9);
+        assert_eq(unsafe_mul(9_000_000_000, 1), 9);
+    }
+
+    #[test]
+    #[expected_failure(abort_code = EUnderflow)]
+    fun test_mul_underflow() {
+        safe_mul(99_999_999, 1);
+    }
+
+    #[test]
+    #[expected_failure(abort_code = EUnderflow)]
+    fun test_mul_round_check_underflow() {
+        safe_mul_round(99_999_999, 1);
+    }
+
+    #[test]
+    fun test_mul_round() {
+        let (is_round, result) = unsafe_mul_round(1_000_000_000, 1);
+        assert_eq(is_round, false);
+        assert_eq(result, 1);
+        (is_round, result) = unsafe_mul_round(9_999_999_999, 1);
+        assert_eq(is_round, true);
+        assert_eq(result, 9);
+    }
+
+    #[test]
+    fun test_div_round() {
+        let (is_round, result) = unsafe_div_round(1, 1_000_000_000);
+        assert_eq(is_round, false);
+        assert_eq(result, 1);
+        (is_round, result) = unsafe_div_round(1, 9_999_999_999);
+        assert_eq(is_round, true);
+        assert_eq(result, 0);
+        (is_round, result) = unsafe_div_round(1, 999_999_999);
+        assert_eq(is_round, true);
+        assert_eq(result, 1);
+    }
+
+    #[test]
+    #[expected_failure(abort_code = EUnderflow)]
+    fun test_div_round_check_underflow() {
+        safe_div_round(1, 1_000_000_001);
+    }
+
+    const DEPRECATED: u64 = 0;
+
+    // === Deprecated ===
+    /// Deprecated since v1.2.0, use `safe_mul` instead.
+    public fun mul(_x: u64, _y: u64): u64 {
```

Comment:
Let's delete `safe_mul` and keep this to avoid bloat?

Reply:
Sorry, could you elaborate? Do you mean by we change all the safe_mul function name to mul? and deprecate the safe_mul?

Reply:
Yeah, I don't fully understand what you mean to express here either.

Reply:
What I mean is:
- Before this PR, there was a function called `mul`. It implemented unsafe multiplication
- This PR introduces functions called `unsafe_mul` and `safe_mul` that do the expected thing
- Instead, this PR should change the body of `mul` to implement safe multiplication, and remove `unsafe_mul`